community/detectors/gitlab_cve_2021_22205/src/main/java/com/google/tsunami/plugins/detectors/cves/cve202122205/Cve202122205VulnDetector.java

```diff
@@ -0,0 +1,244 @@
+          .append("http://")
+          .append(toUriAuthority(networkService.getNetworkEndpoint()))
+          .append("/");
+    }
+    targetUrlBuilder.append(USER_SIGN_PATH);
+    return targetUrlBuilder.toString();
+  }
+
+  private static String buildTargetVulUrl(NetworkService networkService) {
+    StringBuilder targetUrlBuilder = new StringBuilder();
+    if (NetworkServiceUtils.isWebService(networkService)) {
+      targetUrlBuilder.append(NetworkServiceUtils.buildWebApplicationRootUrl(networkService));
+    } else {
+      targetUrlBuilder
+          .append("http://")
+          .append(toUriAuthority(networkService.getNetworkEndpoint()))
+          .append("/");
+    }
+    targetUrlBuilder.append(VUL_PATH);
+    return targetUrlBuilder.toString();
+  }
+
+  @Override
+  public DetectionReportList detect(
+      TargetInfo targetInfo, ImmutableList<NetworkService> matchedServices) {
+    logger.atInfo().log("CVE-2921-22205 starts detecting.");
+
+    return DetectionReportList.newBuilder()
+        .addAllDetectionReports(
+            matchedServices.stream()
+                .filter(Cve202122205VulnDetector::isWebServiceOrUnknownService)
+                .filter(this::isServiceVulnerable)
+                .map(networkService -> buildDetectionReport(targetInfo, networkService))
+                .collect(toImmutableList()))
+        .build();
+  }
+
+  private String getCsrfToken(NetworkService networkService) {
+    String targetUserSignUrl = buildTargetUserSignUrl(networkService);
+    String csrfToken = null;
+    try {
+      HttpResponse httpResponse =
+          httpClient.send(get(targetUserSignUrl).withEmptyHeaders().build(), networkService);
+      if (httpResponse.status().code() == 200) {
+        Matcher csrfTokenMatcher = CSRF_TOKEN_PATTERN.matcher(httpResponse.bodyString().get());
+        if (csrfTokenMatcher.find()) {
+          csrfToken = csrfTokenMatcher.group(1);
+        }
+      }
+    } catch (IOException e) {
+      logger.atWarning().withCause(e).log("Request to target %s failed", networkService);
+      return null;
+    }
+    return csrfToken;
+  }
+
+  private boolean isServiceVulnerable(NetworkService networkService) {
+    String targetVulUrl = buildTargetVulUrl(networkService);
+    String csrfToken = getCsrfToken(networkService);
+    if (csrfToken == null) {
+      logger.atWarning().log("Get %s csrf_token failed", networkService);
+      return false;
+    }
+    try {
+      byte[] payload = Hex.decodeHex(POST_DATA);
+
+      HttpResponse httpResponse =
+          httpClient.send(
+              post(targetVulUrl)
+                  .setHeaders(
+                      HttpHeaders.builder()
+                          .addHeader(
+                              CONTENT_TYPE,
+                              "multipart/form-data; boundary=----WebKitFormBoundaryIMv3"
+                                  + "mxRg59TkFSX5")
+                          .addHeader("X-CSRF-Token", csrfToken)
```

Comment:
You can use the new `addHeader(String name, String value, boolean canonicalize)` method to disable canonicalization.

Reply:
ok